# Introduction
##### This project is about analyzing used cars data from a German web portal

In [2]:
import pandas as pd
import numpy as np

In [3]:
autos = pd.read_csv('autos.csv', encoding='Latin-1')

In [4]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

In [5]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [6]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [7]:
autos.rename(columns = {'dateCrawled':'date_crawled', 
                        'name':'name', 
                        'seller':'seller', 
                        'offerType':'offer_type', 
                        'price':'price', 
                        'abtest':'abtest',
                        'vehicleType': 'vehicle_type', 
                        'yearOfRegistration':'registration_year', 
                        'gearbox':'gearbox', 
                        'powerPS':'powerPS', 
                        'model':'model',
                        'odometer':'odometer', 
                        'monthOfRegistration':'registration_month', 
                        'fuelType':'fuel_type', 
                        'brand':'brand',       
                        'notRepairedDamage':'unrepaired_damage', 
                        'dateCreated':'ad_created', 
                        'nrOfPictures':'nr_of_pictures', 
                        'postalCode':'postal_code',
                        'lastSeen':'last_seen'}, inplace = True)

In [8]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  powerPS   model   odometer  registration_month fuel_type  \
0    manuell      158  andere  150,000km                   3       lpg   
1  automatik      286     7er  150,000km                   6    benzin   
2    manuell      102    golf   70,000km                   7    benzin   
3  automatik       71  fortwo   70,000km                   6    benzin   
4    manuell        0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

##### Column names renamed and case changed to snake case for better handling and understandability

In [9]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-19 17:36:18  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox       powerPS  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

### Observations and Actions
#### We see that 'seller' and 'offer_type' columns have only one value throughout the dataset indicating that they are not helpful for analysis
#### Registration year and ad_created columns need some investigation as they have NaN values 

In [10]:
autos['price'].describe()

count     50000
unique     2357
top          $0
freq       1421
Name: price, dtype: object

In [11]:
autos['price'] = autos['price'].str.replace('$','')
autos['price'] = autos['price'].str.replace(',','')
autos['price'] = autos['price'].astype(int)
autos['price'].head()

0    5000
1    8500
2    8990
3    4350
4    1350
Name: price, dtype: int64

In [12]:
autos['odometer'] = autos['odometer'].str.replace('km','')
autos['odometer'] = autos['odometer'].str.replace(',','')
autos['odometer'] = autos['odometer'].astype(int)
autos['odometer'].head()

0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer, dtype: int64

In [13]:
autos.rename(columns={'odometer':'odometer_km'}, inplace=True)
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'powerPS', 'model',
       'odometer_km', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

In [14]:
autos['odometer_km'].value_counts()

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64

In [15]:
autos[autos['price'].between(0,500000)].shape

(49986, 20)

In [16]:
autos.loc[autos['price']>100000,'brand'].value_counts()

porsche           26
sonstige_autos     8
bmw                4
mercedes_benz      4
ford               3
volkswagen         3
opel               1
citroen            1
volvo              1
fiat               1
audi               1
Name: brand, dtype: int64

In [17]:
autos.loc[autos['brand']=='porsche','price'].value_counts(bins=5)

(-350.001, 70000.0]     243
(70000.0, 140000.0]      38
(140000.0, 210000.0]      8
(280000.0, 350000.0]      3
(210000.0, 280000.0]      2
Name: price, dtype: int64

### Observations and actions
#### Upon exploring the odometer and price columns, it can be seen that odometer seemed reasonable whereas few of the price values are too high to be true. After exploring the price column, we can take a cut off at 500,000 dollars and disregard the values beyond that value which come down to 14. This seems reasonable as we are liberal in disregarding facts as some of the Porsche cars are really costly.

In [18]:
autos = autos[autos['price'].between(0,500000)]

In [19]:
autos.shape

(49986, 20)

In [20]:
autos['date_crawled'].str[:10].value_counts(normalize = True, dropna = False)

2016-04-03    0.038691
2016-03-20    0.037831
2016-03-21    0.037490
2016-03-12    0.036770
2016-03-14    0.036630
2016-04-04    0.036490
2016-03-07    0.035970
2016-04-02    0.035410
2016-03-19    0.034910
2016-03-28    0.034850
2016-03-29    0.034150
2016-03-15    0.033990
2016-04-01    0.033809
2016-03-30    0.033629
2016-03-08    0.033269
2016-03-09    0.033209
2016-03-22    0.032909
2016-03-26    0.032489
2016-03-11    0.032489
2016-03-23    0.032389
2016-03-10    0.032129
2016-03-31    0.031909
2016-03-25    0.031749
2016-03-17    0.031509
2016-03-27    0.031049
2016-03-16    0.029508
2016-03-24    0.029108
2016-03-05    0.025387
2016-03-13    0.015564
2016-03-06    0.013944
2016-04-05    0.013104
2016-03-18    0.013064
2016-04-06    0.003181
2016-04-07    0.001420
Name: date_crawled, dtype: float64

In [21]:
autos['ad_created'].str[:10].value_counts(normalize = True, dropna = False)

2016-04-03    0.038931
2016-03-20    0.037871
2016-03-21    0.037691
2016-04-04    0.036850
2016-03-12    0.036610
2016-03-14    0.035230
2016-04-02    0.035090
2016-03-28    0.034970
2016-03-07    0.034750
2016-03-29    0.034110
2016-03-19    0.033849
2016-04-01    0.033809
2016-03-15    0.033749
2016-03-30    0.033449
2016-03-08    0.033309
2016-03-09    0.033229
2016-03-11    0.032789
2016-03-22    0.032769
2016-03-26    0.032569
2016-03-23    0.032189
2016-03-31    0.031909
2016-03-25    0.031889
2016-03-10    0.031869
2016-03-17    0.031189
2016-03-27    0.030909
2016-03-16    0.030008
2016-03-24    0.029088
2016-03-05    0.023046
2016-03-13    0.016925
2016-03-06    0.015124
                ...   
2016-01-27    0.000060
2016-01-10    0.000040
2016-02-14    0.000040
2016-02-20    0.000040
2016-02-18    0.000040
2016-02-02    0.000040
2016-02-05    0.000040
2016-02-24    0.000040
2016-02-26    0.000040
2016-02-09    0.000040
2016-01-13    0.000020
2016-02-22    0.000020
2016-02-11 

In [22]:
autos['last_seen'].str[:10].value_counts(normalize = True, dropna = False).sort_index()

2016-03-05    0.001080
2016-03-06    0.004421
2016-03-07    0.005362
2016-03-08    0.007582
2016-03-09    0.009843
2016-03-10    0.010763
2016-03-11    0.012524
2016-03-12    0.023807
2016-03-13    0.008983
2016-03-14    0.012804
2016-03-15    0.015884
2016-03-16    0.016445
2016-03-17    0.027928
2016-03-18    0.007422
2016-03-19    0.015744
2016-03-20    0.020706
2016-03-21    0.020726
2016-03-22    0.021586
2016-03-23    0.018585
2016-03-24    0.019565
2016-03-25    0.019205
2016-03-26    0.016965
2016-03-27    0.016024
2016-03-28    0.020846
2016-03-29    0.022326
2016-03-30    0.024847
2016-03-31    0.023827
2016-04-01    0.023106
2016-04-02    0.024887
2016-04-03    0.025367
2016-04-04    0.024627
2016-04-05    0.124275
2016-04-06    0.220982
2016-04-07    0.130957
Name: last_seen, dtype: float64

In [23]:
autos['registration_year'].describe()

count    49986.000000
mean      2005.075721
std        105.727161
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

#### Observations and Actions
##### Registration year seems to be having some irrelevant values like 9999 and 1000. These ought to be removed

In [24]:
autos.loc[autos['registration_year'].between(1900,1950),'registration_year'].value_counts()

1910    9
1937    4
1950    3
1941    2
1934    2
1948    1
1943    1
1939    1
1938    1
1931    1
1929    1
1927    1
Name: registration_year, dtype: int64

##### Cutoff year for registrations could be taken as 1910 assuming such old cars are indeed still being used and listed on the portal. 

In [25]:
autos = autos[autos['registration_year'].between(1910,2016)]
autos['registration_year'].value_counts(normalize = True).sort_index()

1910    0.000187
1927    0.000021
1929    0.000021
1931    0.000021
1934    0.000042
1937    0.000083
1938    0.000021
1939    0.000021
1941    0.000042
1943    0.000021
1948    0.000021
1950    0.000062
1951    0.000042
1952    0.000021
1953    0.000021
1954    0.000042
1955    0.000042
1956    0.000104
1957    0.000042
1958    0.000083
1959    0.000146
1960    0.000687
1961    0.000125
1962    0.000083
1963    0.000187
1964    0.000250
1965    0.000354
1966    0.000458
1967    0.000562
1968    0.000541
          ...   
1987    0.001562
1988    0.002957
1989    0.003770
1990    0.008226
1991    0.007414
1992    0.008122
1993    0.009268
1994    0.013745
1995    0.027324
1996    0.030073
1997    0.042236
1998    0.051087
1999    0.062438
2000    0.069852
2001    0.056273
2002    0.052753
2003    0.056794
2004    0.057002
2005    0.062792
2006    0.056377
2007    0.047984
2008    0.046464
2009    0.043673
2010    0.033260
2011    0.034030
2012    0.027553
2013    0.016786
2014    0.0138

In [26]:
autos['brand'].value_counts(normalize = True)[0:10]

volkswagen       0.212117
bmw              0.110026
opel             0.108172
mercedes_benz    0.095364
audi             0.086409
ford             0.069768
renault          0.047359
peugeot          0.029532
fiat             0.025866
seat             0.018181
Name: brand, dtype: float64

### Observations and Actions
##### We can see that Volkswagen has the max share of 21%. We can explore the top 10 brands by share of cars as they accounted for more than 80% of the cars in the list.

In [27]:
autos['brand'].value_counts()[0:10].index

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat'],
      dtype='object')

In [28]:
top10_brands_price = {}
for car in autos['brand'].value_counts()[0:10].index:
    top10_brands_price[car] = autos.loc[autos['brand']==car,'price'].mean()

top10_brands_mileage = {}
for car in autos['brand'].value_counts()[0:10].index:
    top10_brands_mileage[car] = autos.loc[autos['brand']==car,'odometer_km'].mean()

top10_brands_yrs_used = {}
for car in autos['brand'].value_counts()[0:10].index:
    selected_brand = autos[autos['brand']==car]
    years = selected_brand['date_crawled'].str[:4].astype(int)-selected_brand['registration_year']
    top10_brands_yrs_used[car] = years.mean() 

df = pd.DataFrame(pd.Series(top10_brands_price), columns = ['mean_price'])
df['mean_mileage'] = pd.Series(top10_brands_mileage)
df['mean_yrs_used'] = pd.Series(top10_brands_yrs_used)

mean_price   mean_mileage  mean_yrs_used
audi           9093.650036  129287.780188      11.929622
bmw            8102.536248  132431.383684      13.066629
fiat           2711.801127  116553.945250      12.937198
ford           3652.095224  124068.656716      13.340299
mercedes_benz  8485.239572  130856.082114      14.019655
opel           2876.716404  129223.142087      13.859068
peugeot        3039.468265  127136.812412      12.275740
renault        2395.416447  128183.817062      13.176781
seat           4296.492554  121563.573883      11.586483
volkswagen     5231.081983  128724.104075      13.640157

In [31]:
np.round(df,2).sort_values('mean_price', ascending = False)

mean_price  mean_mileage  mean_yrs_used
audi              9093.65     129287.78          11.93
mercedes_benz     8485.24     130856.08          14.02
bmw               8102.54     132431.38          13.07
volkswagen        5231.08     128724.10          13.64
seat              4296.49     121563.57          11.59
ford              3652.10     124068.66          13.34
peugeot           3039.47     127136.81          12.28
opel              2876.72     129223.14          13.86
fiat              2711.80     116553.95          12.94
renault           2395.42     128183.82          13.18

## Observations 

#### We can see that the mileage is more or less same and even the no